cypress-io/cypress
# feat: support React 18

- Closes https://github.com/cypress-io/cypress/issues/21381
- Closes https://github.com/cypress-io/cypress/issues/21946

### User facing changelog


Add support for React 18 to `npm/react`.

### Additional details


- updated `peerDependencies` to allow installing `@cypress/react` with `react@18`
- some specific code in vite and webpack dev servers to account for the potentially not available `react-dom/client` code

The problem and solution is described in detail in the README [in this repo](https://github.com/lmiller1990/es-modules-example) which is an example project used for an internal presentation at Cypress.

### Steps to test


- Build binary (local should be fine too, but binary is more "production like")
- Try using Cypress CT with a React 18 project (using new `react-dom/client` API) 
  -  For example 
    - webpack: `yarn create react-app my-app` 
    - vite: `yarn create vite --template react my-vite-react-app`
  - Point Cypress at your app
  - Configure React using Launchpad
  - Should work, no errors in console, etc.
- No warning/error in console about deprecated React 18 API etc.

### How has the user experience changed?


### PR Tasks
